### ISSUE TITLE
ScriptUserDefinition GetUserData() caches first persons permission

### ISSUE THREAD

**Issue by abelal83** [volkanceylan/Serenity#1248]
Odd problem...

I've modified GetUserData to retrieve permissions from Active Directory as below

```
public ScriptUserDefinition GetUserData()
        {

            PrincipalContext yourDomain = new PrincipalContext(ContextType.Domain, Authorization.Username.Split('\\')[0]);
            UserPrincipal userPrincipal = UserPrincipal.FindByIdentity(yourDomain, Authorization.Username.Split('\\')[1]);

            var result = new ScriptUserDefinition();
            var user = Authorization.UserDefinition as UserDefinition;

            if (user == null)
            {
                result.Permissions = new Dictionary<string, bool>();
                return result;
            }

            result.Username = Authorization.Username; //user.Username;
            result.DisplayName = user.DisplayName;

            result.Permissions = TwoLevelCache.GetLocalStoreOnly("ScriptUserPermissions:" + user.Id, TimeSpan.Zero,
                //UserPermissionRow.Fields.GenerationKey, () =>
                "ScriptUserPermissions:" + user.Id, () =>
                {
                    var permissions = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);

                    //if (permissionsUsedFromScript == null)
                    if (permissionsUsedFromScript.Count == 0)
                    {
                        //permissionsUsedFromScript = new UserPermissionRepository().ListPermissionKeys().Entities
                        //    .Where(permissionKey =>
                        //    {
                        //        // this sends permission information for all permission keys to client side.
                        //        // if you don't need all of them to be available from script, filter them here.
                        //        // this is recommended for security / performance reasons...
                        //        return true;
                        //    }).ToArray();

                        if (userPrincipal != null)
                        {
                            PrincipalSearchResult<Principal> groups = userPrincipal.GetAuthorizationGroups();

                            foreach (Principal p in groups)
                            {
                                // make sure to add only group principals
                                if (p is GroupPrincipal)
                                {
                                    try
                                    {
                                        var account = p.Sid.Translate(typeof(NTAccount));

                                        permissionsUsedFromScript.Add(account.Value);
                                    }
                                    catch
                                    {
                                        
                                    }
                                }
                            }                            
                        }
                    }
                    
                    foreach (var permissionKey in permissionsUsedFromScript)
                    {
                        //if (Authorization.HasPermission(permissionKey))
                            permissions[permissionKey] = true;
                    }

                    TwoLevelCache.Remove("ScriptUserPermissions:" + user.Id);
                    return permissions;
                });

            return result;
        }
```

the problem I seem to be getting now is the very first person who logs on, their permission seems to be used for everyone else until I reset iis or wait some time. It sounds like a cache but as a test I've explicitly removed the case at the end. Any ideas?

**Comment by volkanceylan** [comment]
This is only a method used to generate data for client side permissions. Are you sure you didn't modify PermissionService etc. ?

**Comment by abelal83** [comment]
yep understand it's for client side permission only.

I'm using the below code in xyzDialog.ts to enable\disable an item based on Active Directory permission
```
        afterLoadEntity() {
            super.afterLoadEntity();
            if (!Authorization.hasPermission("mydomain\\team-deskside-ldn")) {
                this.form.Permanent.getGridField().toggle(false);
            }
        }
```

As I debugged the code I noticed this..
The first time run serenity the calls to methods are like so
 !Authorization.hasPermission("mydomain\\team-deskside-ldn") -> Q.getRemoteData('UserData') -> UserDefinition GetFirst()

on subsequent runs the calls to method are like this 
!Authorization.hasPermission("mydomain\\team-deskside-ldn") -> Q.getRemoteData('UserData') 

it never calls GetFirst again which means it doesn't build the permissions for any other users.

What decides when GetFirst is called?

**Comment by volkanceylan** [comment]
Users are cached by their name. So no need to call GetFirst if same user is requested second time. I guess you don't have negative expiration value on your GetUserData method. e.g. [NonAction, DataScript("UserData", CacheDuration = -1)]

Or, somehow Authorization.UserId or Authorization.Username doesn't change etc.

**Comment by abelal83** [comment]
I figured it out, because I had changed permissionUsedFromScript to a List object it was being reused, this contained the permission from previous user.

```
        //private static string[] permissionsUsedFromScript;
        private static List<string> permissionsUsedFromScript = new List<string>();
```
I amended code to include

`permissionsUsedFromScript.Clear();`

this resolved the issue. I now have a fully integrated Active Directory serene environment using single sign on! 

I just need to improve my implementation of GetUserData() as querying AD is slooooow compared to database, of course that's not a serenity issue.

Thanks as usual for the help Volkan.
